[JSC] Add support for new Set.prototype methods

kmiller68 opened the pull request:
#### ff30d6f8aba3d3638bd80f5ccea91d9e03535598
<pre>
[JSC] Add support for new Set.prototype methods
<a href="https://bugs.webkit.org/show_bug.cgi?id=251510">https://bugs.webkit.org/show_bug.cgi?id=251510</a>

Reviewed by Yusuke Suzuki.

This patch adds support for the <a href="https://github.com/tc39/proposal-set-methods">https://github.com/tc39/proposal-set-methods</a> proposal, which is currently at stage 3.
In order to unobservably make a copy of a set this patch adds a new setClone intrisic, otherwise everything else is just
normal builtins code. There is one outstanding issue that intersection is supposed to return a Set with the same iteration
order as the this value. Given our implementation of iteration order being a linked list that&apos;s not efficient. Pending spec discussion
this implementation just skips that step. See: <a href="https://bugs.webkit.org/show_bug.cgi?id=251869">https://bugs.webkit.org/show_bug.cgi?id=251869</a>

* JSTests/stress/set-prototype-intersection.js: Added.
(assert):
(assertArrayContent):
(catch):
(try.set1.intersection.has):
(try.set1.intersection.keys):
(assertArrayContent.Array.from.set1.intersection.has):
(assertArrayContent.Array.from.set1.intersection.keys):
(assertArrayContent.Array.from.set4.intersection.has):
(assertArrayContent.Array.from.set4.intersection.keys):
* JSTests/stress/set-prototype-isDisjointfrom.js: Added.
(assert):
(assertArrayContent):
(catch):
(try.set1.isDisjointFrom.has):
(try.set1.isDisjointFrom.keys):
(assert.set1.isDisjointFrom.has):
(assert.set1.isDisjointFrom.keys):
(assert.set4.isDisjointFrom.has):
(assert.set4.isDisjointFrom.keys):
(assert.set6.isDisjointFrom.has):
(assert.set6.isDisjointFrom.keys):
* JSTests/stress/set-prototype-isSubsetOf.js: Added.
(assert):
(assertArrayContent):
(catch):
(try.set1.isSubsetOf.has):
(try.set1.isSubsetOf.keys):
(assert.set1.isSubsetOf.has):
(assert.set1.isSubsetOf.keys):
(assert.set4.isSubsetOf.has):
(assert.set4.isSubsetOf.keys):
* JSTests/stress/set-prototype-isSupersetOf.js: Added.
(assert):
(assertArrayContent):
(catch):
(try.set1.isSupersetOf.has):
(try.set1.isSupersetOf.keys):
(assert.set1.isSupersetOf.has):
(assert.set1.isSupersetOf.keys):
(assert.set4.isSupersetOf.has):
(assert.set4.isSupersetOf.keys):
* JSTests/stress/set-prototype-symmetricDifference.js: Added.
(assert):
(assertArrayContent):
(catch):
(try.set1.symmetricDifference.has):
(try.set1.symmetricDifference.keys):
(assertArrayContent.Array.from.set1.symmetricDifference.has):
(assertArrayContent.Array.from.set1.symmetricDifference.keys):
(assertArrayContent.Array.from.set4.symmetricDifference.has):
(assertArrayContent.Array.from.set4.symmetricDifference.keys):
* JSTests/stress/set-prototype-union.js: Added.
(assert):
(assertArrayContent):
(catch):
(try.set1.union.has):
(try.set1.union.keys):
(assertArrayContent.Array.from.set1.union.has):
(assertArrayContent.Array.from.set1.union.keys):
(assertArrayContent.Array.from.set4.union.has):
(assertArrayContent.Array.from.set4.union.keys):
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/SetPrototype.js:
(union.wrapper.iterator):
(union):
(intersection.else.wrapper.iterator):
(intersection):
(symmetricDifference.wrapper.iterator):
(symmetricDifference):
(isSubsetOf):
(isSupersetOf.wrapper.iterator):
(isSupersetOf):
(isDisjointFrom.else.wrapper.iterator):
(isDisjointFrom):
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/SetConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/SetConstructor.h:
* Source/JavaScriptCore/runtime/SetPrototype.cpp:
(JSC::SetPrototype::finishCreation):

Canonical link: <a href="https://commits.webkit.org/260033@main">https://commits.webkit.org/260033@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/03770e32c2f5fbbce1f9980be6fa005265b0d1d3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106779 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15793 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39565 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115968 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115545 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17297 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/7049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98974 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112548 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13117 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96098 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40712 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95000 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/27750 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82444 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96317 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8986 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/29105 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/95787 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/6920 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9551 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/7049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/30698 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15155 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48648 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/104538 "Built successfully") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/6929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11074 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25890 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3740 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/builtins/SetPrototype.js:
```diff
@@ -44,3 +44,254 @@ function forEach(callback /*, thisArg */)
+            if (has.@call(other, key))
+                result.@add(key);
+        } while (true);
+    } else {
```
Probably, we should have a Set fast path for other case. (iterating things like Set.forEach).

Jarred-Sumner replied:
it would be really cool if there was a way for C++-implemented classes to expose a fast path for iterators more generally. Like if the Iterable object had an optional private property to call a function that returns or inserts the rest of the results as a JSArray or JSObject. `Array.from()`, `Object.fromEntries()`,  `[...spread]`, could use that automatically. Maybe a generic template with C++ iterators could be used to generate the fast path so that consumers in WebCore/elsewhere wouldn't need to bother with implementing the fast path function themselves

This would speed up iterators in Headers, the FileSystem Access API and anywhere else
```js
    const headers = new Headers({
      "Last-Modified": new Date().toUTCString(),
      ETag: `W/abcdef`,
    });
   
   const pairs = [...headers.entries()]
   const keys = [...headers.keys()]
   const values = [...headers.keys()]
```

kmiller68 replied:
Yeah, I could see doing something. Idk if we'd expose a way to say what the effects of the callback are to the DFG/FTL though. I could see most people getting that wrong and causing weird bugs/crashes/security issues for themselves.

That all said, I think we should add tuples to the DFG first so we can have a generic inline caching version of the iterator operations first. Otherwise the polymorphic case starts to get significantly slower than in current code due to DFG getting confused about the control flow of the program.

kmiller68 replied:
> Probably, we should have a Set fast path for other case. (iterating things like Set.forEach).

Iterating via the wrapper should be just as fast in the DFG/FTL because we should inline all the calls (maybe a bit worse in the DFG because you have to allocate the wrapper).

Constellation commented on Source/JavaScriptCore/builtins/SetPrototype.js:
```diff
@@ -44,3 +44,254 @@ function forEach(callback /*, thisArg */)
+    };
+
+    var result = @setClone(this);
+    for (var key of wrapper)
+        result.@add(key);
```
Ditto.

Constellation commented on Source/JavaScriptCore/builtins/SetPrototype.js:
```diff
@@ -44,3 +44,254 @@ function forEach(callback /*, thisArg */)
+    };
+
+    var result = @setClone(this);
+    for (var key of wrapper) {
+        if (result.@has(key))
+            result.@delete(key);
+        else
+            result.@add(key);
+    }
```
Ditto.

Constellation commented on Source/JavaScriptCore/builtins/SetPrototype.js:
```diff
@@ -44,3 +44,254 @@ function forEach(callback /*, thisArg */)
+    if (this.@size < size)
+        return false;
+
+    var iterator = keys.@call(other, keys);
+    var wrapper = {
+        @@iterator: function () { return iterator; }
+    };
+
+    for (var key of wrapper) {
+        if (!this.@has(key))
+            return false;
+    }
```
Ditto.

Constellation commented on Source/JavaScriptCore/builtins/SetPrototype.js:
```diff
@@ -44,3 +44,254 @@ function forEach(callback /*, thisArg */)
+                return false;
+        } while (true);
+    } else {
+        var iterator = keys.@call(other, keys);
+        var wrapper = {
+            @@iterator: function () { return iterator; }
+        };
+
+        for (var key of wrapper) {
+            if (this.@has(key))
+                return false;
+        }
```
Ditto.

Constellation reviewed (approved):
r=me with comments. I think we need to add fast path before enabling this.

rkirsling commented on Source/JavaScriptCore/builtins/SetPrototype.js:
```diff
@@ -44,3 +44,254 @@ function forEach(callback /*, thisArg */)
+            @@iterator: function () { return iterator; }
+        };
+
+        for (var key of wrapper) {
+            if (this.@has(key))
+                result.@add(key);
+        }
```
With the current spec, you'll actually need to add to a temporary set and then iterate over the receiver in order to ensure that the elements remain in their original order. This was going to be my comment:
```
// The idea of the current spec is that:
// 1. We want to have the output elements stay in their original order.
// 2. We don't want to perform a greater number of user-observable has() calls on the argument set
//    than the number of elements in it.
//
// So, we iterate over the argument set when it's the smaller one, but the underlying assumption is
// that we *won't* need to iterate over the receiver set in order to sort the keys. In SM & V8,
// "deterministic hash tables" allow the ordering of two elements to be checked by a pointer comparison,
// but our implementation (with a linked list of buckets) does not allow for this.
// Accordingly, the spec may change here.
```
I was discussing this with Kevin yesterday, as I [shared on Slack](https://webkit.slack.com/archives/CTV4FGWF4/p1675219659716579?thread_ts=1675150621.902089&cid=CTV4FGWF4), and it sounds like he'll reraise this to committee in the near future.

kmiller68 replied:
Oof, I guess I missed that step when I read the spec. I agree that's not great for our implementation... For now I'll just put a fixme in the code and in the option.

kmiller68 committed:
[JSC] Add support for new Set.prototype methods
https://bugs.webkit.org/show_bug.cgi?id=251510

Reviewed by Yusuke Suzuki.

This patch adds support for the https://github.com/tc39/proposal-set-methods proposal, which is currently at stage 3.
In order to unobservably make a copy of a set this patch adds a new setClone intrisic, otherwise everything else is just
normal builtins code. There is one outstanding issue that intersection is supposed to return a Set with the same iteration
order as the this value. Given our implementation of iteration order being a linked list that's not efficient. Pending spec discussion
this implementation just skips that step. See: https://bugs.webkit.org/show_bug.cgi?id=251869

* JSTests/stress/set-prototype-intersection.js: Added.
(assert):
(assertArrayContent):
(catch):
(try.set1.intersection.has):
(try.set1.intersection.keys):
(assertArrayContent.Array.from.set1.intersection.has):
(assertArrayContent.Array.from.set1.intersection.keys):
(assertArrayContent.Array.from.set4.intersection.has):
(assertArrayContent.Array.from.set4.intersection.keys):
* JSTests/stress/set-prototype-isDisjointfrom.js: Added.
(assert):
(assertArrayContent):
(catch):
(try.set1.isDisjointFrom.has):
(try.set1.isDisjointFrom.keys):
(assert.set1.isDisjointFrom.has):
(assert.set1.isDisjointFrom.keys):
(assert.set4.isDisjointFrom.has):
(assert.set4.isDisjointFrom.keys):
(assert.set6.isDisjointFrom.has):
(assert.set6.isDisjointFrom.keys):
* JSTests/stress/set-prototype-isSubsetOf.js: Added.
(assert):
(assertArrayContent):
(catch):
(try.set1.isSubsetOf.has):
(try.set1.isSubsetOf.keys):
(assert.set1.isSubsetOf.has):
(assert.set1.isSubsetOf.keys):
(assert.set4.isSubsetOf.has):
(assert.set4.isSubsetOf.keys):
* JSTests/stress/set-prototype-isSupersetOf.js: Added.
(assert):
(assertArrayContent):
(catch):
(try.set1.isSupersetOf.has):
(try.set1.isSupersetOf.keys):
(assert.set1.isSupersetOf.has):
(assert.set1.isSupersetOf.keys):
(assert.set4.isSupersetOf.has):
(assert.set4.isSupersetOf.keys):
* JSTests/stress/set-prototype-symmetricDifference.js: Added.
(assert):
(assertArrayContent):
(catch):
(try.set1.symmetricDifference.has):
(try.set1.symmetricDifference.keys):
(assertArrayContent.Array.from.set1.symmetricDifference.has):
(assertArrayContent.Array.from.set1.symmetricDifference.keys):
(assertArrayContent.Array.from.set4.symmetricDifference.has):
(assertArrayContent.Array.from.set4.symmetricDifference.keys):
* JSTests/stress/set-prototype-union.js: Added.
(assert):
(assertArrayContent):
(catch):
(try.set1.union.has):
(try.set1.union.keys):
(assertArrayContent.Array.from.set1.union.has):
(assertArrayContent.Array.from.set1.union.keys):
(assertArrayContent.Array.from.set4.union.has):
(assertArrayContent.Array.from.set4.union.keys):
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/SetPrototype.js:
(union.wrapper.iterator):
(union):
(intersection.else.wrapper.iterator):
(intersection):
(symmetricDifference.wrapper.iterator):
(symmetricDifference):
(isSubsetOf):
(isSupersetOf.wrapper.iterator):
(isSupersetOf):
(isDisjointFrom.else.wrapper.iterator):
(isDisjointFrom):
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/SetConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/SetConstructor.h:
* Source/JavaScriptCore/runtime/SetPrototype.cpp:
(JSC::SetPrototype::finishCreation):

Canonical link: https://commits.webkit.org/260033@main
